### PR TITLE
Adding Graphormer layer

### LINKED ISSUE #251: implement Graphormer
https://proceedings.neurips.cc/paper/2021/file/f1c1592588411002af340cbaedd6fc33-Paper.pdf

### PR DESCRIPTION
**Pull request by 5hv5hvnk**
Continued from   #265
fixes #251 
To do:
- [ ] struct based layers
- [ ] doc-string
- [ ] example pipeline
- [ ] unit tests
- [ ] pass build tests
- [ ] fixed conflicts

### CODE REVIEW

File: src/layers/conv.jl [hunk 1]
```diff
@@ -1688,3 +1688,83 @@ function Base.show(io::IO, l::TransformerConv)
...
+    negative_slope::T
+    channel::Pair{NTuple{2, Int}, Int}
+    heads::Int
+    concat::Bool
+    add_self_loops::Bool
+end
+
+@functor GraphormerLayer
+
+Flux.trainable(l::GraphormerLayer) = (l.dense_x, l.dense_e, l.bias, l.a)
```
Comment (CarloLucibello) [inline src/layers/conv.jl:1723]:
```suggestion
Flux.trainable(l::GraphormerLayer) = (l.dense_x, l.dense_e, l.bias, l.a)

```

File: src/layers/conv.jl [hunk 2]
```diff
@@ -1688,3 +1688,83 @@ function Base.show(io::IO, l::TransformerConv)
...
+    end
+
+    _, chout = l.channel
+    heads = l.heads
+
+    Wx = l.dense_x(x)
+    Wx = reshape(Wx, chout, heads, :)
+
+    # a hand-written message passing
+    m = apply_edges((xi, xj, e) -> message(l, xi, xj, e), g, Wx, Wx, e)
```
Comment (CarloLucibello) [inline src/layers/conv.jl:1762]:
```suggestion
    m = apply_edges((xi, xj, e) -> message(l, xi, xj, e), g, Wx, Wx, e)
```
where is the message function defined?

Comment (CarloLucibello) [inline src/layers/conv.jl:1762]: where is the message function defined?

**Comment by CarloLucibello** [comment]
This is not capturing the essential feature of the [paper](https://proceedings.neurips.cc/paper/2021/file/f1c1592588411002af340cbaedd6fc33-Paper.pdf):
- The message function is not defined. Since this layer is basically a transformer, the attention runs on the whole graph instead of just the neighborhood of a node. Therefore, we shouldn't use the usual message passing scheme but we can rely on the MultiHeadAttention layer from Flux.
- The bias term of eq. 5 is missing. The phi function should be given at construction time
- As part of the layer, eq. 8 and eq. 9 prescribe layer norm, residual connection, and an MLP. 
- edge features can be considered in a second PR, and should implement the mechanics of eq. 7

**Comment by CarloLucibello** [comment]
Why you checked the tests as done? I see no tests...

File: src/layers/conv.jl (hunk 2 above)
Comment (5hv5hvnk) [reply src/layers/conv.jl:1762]: Apologies I missed writing it I added the comment to go back and add that let me get on to this

**Comment by 5hv5hvnk** [comment]
> Why you checked the tests as done? I see no tests...

I have written some tests you can check [here](https://github.com/5hv5hvnk/GraphNeuralNetworks.jl/blob/6aa987d0dc35c2c84b2d8c3f0f73a87d4883c031/test/layers/conv.jl#L338) 
I don't know why It isn't showing in the files changed let me look into it.

**Comment by 5hv5hvnk** [comment]
> I have written some tests you can check [here](https://github.com/5hv5hvnk/GraphNeuralNetworks.jl/blob/6aa987d0dc35c2c84b2d8c3f0f73a87d4883c031/test/layers/conv.jl#L338) I don't know why It isn't showing in the files changed let me look into it.
@CarloLucibello Hi can you look into the updates and see the tests?

File: src/layers/conv.jl [hunk 3]
```diff
@@ -1688,3 +1688,94 @@ function Base.show(io::IO, l::TransformerConv)
...
+        g = add_self_loops(g)
+    end
+
+    _, chout = l.channel
+    heads = l.heads
+    Wx = l.dense_x(x)
+    Wx = reshape(Wx, chout, heads, :)
+    h = Wx
+    for i in 1:l.phi
+        h = l.MHA(l.LN(h))
```
Comment (CarloLucibello) [inline src/layers/conv.jl:1761]: how is this supposed to work? What are MHA and LN? you should go beyond having a template and produce some working code and some examples

File: src/layers/conv.jl [hunk 4]
```diff
@@ -1688,3 +1688,94 @@ function Base.show(io::IO, l::TransformerConv)
...
+        h = h + Wx
+    end
+    h = reshape(h, size(h, 1), heads * chout)
+    if l.bias !== false
+        h = l.bias(h)
+    end
+    h = l.σ.(h)
+    return h
+end
+function message(l::GraphormerLayer, xi, xj, e)
```
Comment (CarloLucibello) [inline src/layers/conv.jl:1771]: You define this message function but it is not used anywhere

**Comment by CarloLucibello** [comment]
> @CarloLucibello Hi can you look into the updates and see the tests?

can you make the tests part of this PR? it's weird they are somewhere else

**Comment by 5hv5hvnk** [comment]
Hello @CarloLucibello Sorry for delayed response on this I believe we still need this layer let me get on to this.

File: src/layers/conv.jl [hunk 5]
```diff
@@ -1689,3 +1689,114 @@ function Base.show(io::IO, l::TransformerConv)
...
+end
+
+function message(l::GraphormerLayer, xi, xj, e)
+    θ = cat(xi, xj, dims=2)
+    if l.dense_e !== nothing
+        fe = l.dense_e(e)
+        fe = reshape(fe, size(fe, 1), l.heads, :)
+        θ = cat(θ, fe, dims=1)
+    end
+    α = softmax(l.a' * θ)  ## Not sure if I can use it from flux directly
```
Comment (5hv5hvnk) [inline src/layers/conv.jl:1798]: Can you confirm if this is okay?

**Pull request closed without merging**